derived dim

avikchaudhuri opened the pull request:
With the current `Dim`-based dynamic shapes API for export, one can express that shapes of different input shapes must be equal by reusing the same `Dim`. However, non-trivial relationships between such input shapes cannot be expressed.

Recently we are seeing more and more examples of code that require this additional expressibility, e.g., where a pair of shapes might differ by one, or a shape might be double another (or simply even).

This PR introduces the concept of a "derived" `Dim`, i.e., a linear arithmetic expression over a `Dim`. By using a combination of `Dim`s and derived `Dim`s to specify input shapes, the desired relationships can be expressed naturally. E.g., a pair of shapes might be `dim` and `dim + 1`, or `dim` and `2*dim`, or even `2*dim` and `dim + 1`.

We extend the current infrastructure that translates `Dim`s to deprecated `dynamic_dim`-based constraints to work with derived `Dim`s. As usual, we raise constraint violation errors when shape guards cannot be verified given a dynamic shapes spec; suggest fixes; and raise runtime errors when future inputs violate the spec.

Importantly, some guards that used to cause forced specializations in the constraint solver because they were deemed "too complex" now do not do so, because they can now be specified as constraints. Since this was what motivated the introduction of a `disable_constraint_solver` flag to some internal APIs, we may not need that flag any more.

Note that shapes of placeholders in exported programs can now contain symbolic expressions and not just symbols.

Differential Revision: D53254587




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

suo commented on test/export/test_export.py:
```diff
@@ -341,6 +341,175 @@ def forward(self, x, c):
+        dimy = torch.export.Dim("dimy", min=4, max=7)  # doesn't work
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            (
```
maybe use `re.escape` to make this easier to read and right.

suo reviewed (commented):
the frontend bits look reasonable to me but I am not the right person to review the symbolic parts…@ezyang I keep tagging you in on these, is there anyone else

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -811,8 +811,15 @@ class EqualityConstraint(Constraint):
+    Also contains a list of derived equalities, each of which relates a source
+    to an expression over a root. The root can be a source, corresponding to some
+    input shape, or a phantom symbol corresponding to the root Dim of a derived
+    Dim describing some input shape. We track the names of such Dims as well.
```
I'm not sure I understand this explanation of phantom. An example would go a long way here!

avikchaudhuri replied:
A phantom symbol is a symbol that isn't directly the shape of an input tensor, but appears in a symbolic expression that describes such a shape. E.g., an input tensor might have shape `2*s0` where `s0` is not the shape of any input tensor. In that case `s0` is a phantom and doesn't have a source.

ezyang replied:
One thing that concerns me about phantom symbols is that you can easily generate non-instantiatable sizevars. For example, let's say I declare a single tensor input with size `[s0 + s1]`, it is impossible to know what s0 and s1 are, because the phantoms are underconstrained. I wonder what you are doing about this.

avikchaudhuri replied:
The input language for derived dims is deliberately very constrained. One variable, linear expressions of the form `a*x + b`. So this kind of situation cannot happen.

ezyang replied:
I guess (1) people will probably eventually want to added other forms for derived dims, so if this is the invariant we will have to make sure we keep upholding this invariant, and (2) even with linear expressions, it is easy to give an unsolvable equation (since x must be integral), so I will looking out to see where we check for this.

ezyang commented:
@suo we are in an unfortunate place where we have a fairly complicated piece of symbolic reasoning code but it is *only* used by export, so I'm not even that familiar with it (I rubber stamped the initial version that Avik put in, but I never put in the time to understand it in full detail because we basically never hit it in traditional PT2.) So, I am not really the "right" person to review this, but I am probably the right-est person (besides maybe some of our external folks who've also been helping with the non-export symbolic shapes stuff @lezcano, and also @eellison has been helping with the value ranges stuff.)

If we want me to be primary reviewer on this, then some knowledge transfer is going to have to happen, since I don't really understand how the constraint solver currently works and will have to spend some time studying it. I'd also prefer if this can be a bidirectional thing; e.g., @avikchaudhuri commits to reviewing all of the PRs to the regular, online constraint solver (and I can help knowledge transfer in the reverse direction).

avikchaudhuri commented:
> @suo we are in an unfortunate place where we have a fairly complicated piece of symbolic reasoning code but it is _only_ used by export, so I'm not even that familiar with it (I rubber stamped the initial version that Avik put in, but I never put in the time to understand it in full detail because we basically never hit it in traditional PT2.) So, I am not really the "right" person to review this, but I am probably the right-est person (besides maybe some of our external folks who've also been helping with the non-export symbolic shapes stuff @lezcano, and also @eellison has been helping with the value ranges stuff.)
> 
> If we want me to be primary reviewer on this, then some knowledge transfer is going to have to happen, since I don't really understand how the constraint solver currently works and will have to spend some time studying it. I'd also prefer if this can be a bidirectional thing; e.g., @avikchaudhuri commits to reviewing all of the PRs to the regular, online constraint solver (and I can help knowledge transfer in the reverse direction).

I can commit to some knowledge transfer on the constraint solver for export (and by constraint solver, I'm assuming you mean the thing that converts guards into suggested dynamic shape specs for the export front-end). And yes, I'd also love some knowledge transfer in the reverse direction before I can help review PRs related to the online constraint solver (I assume here you mean `evaluate_expr` and friends?).

That said, this PR is not really about the constraint solver per se: it is about addressing the inability to express certain binary relations between dims in the export front-end, and dealing with them completely outside that subsystem.

Now, the export front-end implementation could also be a topic of knowledge transfer. I think @ezyang was familiar with the basic implementation of `dynamic_dim` but then I added equality constraints, then did the `Dim` stuff and its translation into the old system, and now added this derived `Dim` stuff. However, we are getting rid of the old system with `dynamic_dim` soon, and I plan to rip out large parts of the current implementation when that happens so that `Dim` can directly go to symbols. It might be better to spend time discussing how that would look rather than get too deep into this part of the code.

lezcano commented:
Regarding the constraint solver, could you guys add me if there is any call discussing that? Thank you!

ezyang commented on torch/_export/serde/serialize.py:
```diff
@@ -1110,6 +1110,25 @@ def deserialize_sym_int(self, s: SymInt) -> Union[int, torch.SymInt]:
+                                compiler_min=vr.lower,  # type: ignore[arg-type]
+                                compiler_max=vr.upper,  # type: ignore[arg-type]
+                                runtime_min=vr.lower,  # type: ignore[arg-type]
+                                runtime_max=vr.upper  # type: ignore[arg-type]
```
This needs to be rebased past size-agnostic

avikchaudhuri replied:
Oh, I was getting some new errors when I rebased, was wondering why.

ezyang replied:
Just delete the runtime kwargs, should be enough

ezyang commented on torch/export/dynamic_shapes.py:
```diff
@@ -38,6 +42,64 @@ def readable(name, min_, max_):
+    def __add__(cls, other):
+        # e.g., dim + 1
+        assert type(other) is int, f"Expected int, got {type(other)}"
+        return cls._derive(f"({cls.__name__}+{other})", lambda x: x + other)
```
OOC, is there a reason we want to immediately string format as we are constructing _Dim expressions? Seems a bit unusual.

avikchaudhuri replied:
It is. I need to ensure that the name we give to the derived Dim is unique up to structural equality, hence this trick.

avikchaudhuri replied:
I ended up reusing the lambda and the printer for symbolic expressions here in the updated PR, FYI.

avikchaudhuri commented:
> Regarding the constraint solver, could you guys add me if there is any call discussing that? Thank you!

Sent an invite, hope the time works!

avikchaudhuri commented:
Here's some reading material on the existing implementation of the dynamic shapes front-end and the constraint solver for export: https://docs.google.com/document/d/1SpaqdBjWNckJGZj0m3aRSdLMt5fdENp0i6XS_y229og/edit

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2670,7 +2719,7 @@ def produce_guards(
         *,
         input_contexts: Optional[DimList[SymbolicContext]] = None,
-        equalities_inputs: Optional[Set[Tuple[Source, Source]]] = None,
+        equalities_inputs: Optional[EqualityConstraint] = None,
```
Could we get some docs for these inputs?

ezyang replied:
Should EqualityConstraint be perhaps pluralized here? It's a little strange that now there is only one equality constraint, seemingly.

avikchaudhuri replied:
Yeah sure (for both).

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2827,11 +2876,30 @@ def get_symbol(tensor_dim_src):
                 if not concrete_val:
                     raise ConstraintViolationError(
-                        f"{src1.name()} = {self.var_to_val[s1]}"
+                        f"{src1.name()} = {s1.subs(self.var_to_val)}"
```
Odd, is s1 not guaranteed to be a symbol here? Maybe we should rename them (and get_symbol?)

avikchaudhuri replied:
Right, we can no longer assume they are symbols. A placeholder dim might be a symint that holds an expression, something like `s0 + 1`. (It might have started as `s1` but gotten replaced away.)

avikchaudhuri replied:
I'll rename get_symbol for this to be clear.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2827,11 +2876,30 @@ def get_symbol(tensor_dim_src):
+                s1 = get_symbol(src)
+                s2 = root if isinstance(root, sympy.Symbol) else get_symbol(root)
+                s2_ = fn(s2)
+                concrete_val = self.evaluate_expr(sympy.Eq(s1, s2_))
```
evaluate_expr feels like the wrong thing to use here? Because it conventionally will install a guard if we don't know if something is statically the case. Maybe you want `maybe_static_evaluate` instead?

avikchaudhuri replied:
OK, you might be right: if I really want to prove something given the guards, do static evaluation.

avikchaudhuri replied:
So I remembered what this is about, and we do need the `evaluate_expr` here: we want to add a guard if necessary. Specifically, this is checking whether concrete inputs satisfy the specified condition. We're not proving anything.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2827,11 +2876,30 @@ def get_symbol(tensor_dim_src):
+                s2_ = fn(s2)
+                concrete_val = self.evaluate_expr(sympy.Eq(s1, s2_))
+                if not concrete_val:
+                    # we were unable to prove that src = fn(root) holds
```
So, supposing you DO use evaluate_expr, here, not only were you unable to prove that src = fn(root) didn't hold, but in fact, if concrete_val is False, the user has given us inputs for which the constraint was expressly violated!

avikchaudhuri replied:
Yup! So we will emit a `ConstraintViolationError` here.

avikchaudhuri replied:
See above.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2827,11 +2876,30 @@ def get_symbol(tensor_dim_src):
+                concrete_val = self.evaluate_expr(sympy.Eq(s1, s2_))
+                if not concrete_val:
+                    # we were unable to prove that src = fn(root) holds
+                    debug_name = self.var_to_sources[root][0].name() if isinstance(root, sympy.Symbol) else self._debug_name(root)
```
It's weird to me that you can't just unconditionally call _debug_name here

avikchaudhuri replied:
This disjoint union of cases appears in a few places (you'll see a few types in this PR being Union[this, that]), almost missing something like an algebraic datatype here. Or maybe it suggests that the code could be written differently.

avikchaudhuri replied:
Moved around the logic here to make the case analysis clearer.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -845,6 +852,12 @@ def render(self):
+        return any(
+            src == source and fn(symbol) == expr
+            for source, _root, fn in self.derived_equalities
+        )
```
In particular, here, we can potentially have `x = f(y)` and `y = g(z)` and then if we query a guard `x = f(g(z))` we ought to return True when saying that this is_derived.

avikchaudhuri replied:
Agreed.

avikchaudhuri replied:
Yeah, we now have an algorithm for this. See comments. We normalize everything (e.g., sources use their union-find root) and then express everything in terms of normal Dims and check syntactic equality. Normal dims are equal iff they are the same.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1475,6 +1488,43 @@ def solve(self, disable_congruences=True, disable_equivalences=True):
+        return isinstance(expr, sympy.Symbol)
+
+    @classmethod
+    def solve_supported_equivalence(cls, expr, solution):
```
I'm not entirely clear what the signature of this function is, are we solving `expr == solution` or something else?

avikchaudhuri replied:
Using try_solve now.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1475,6 +1488,43 @@ def solve(self, disable_congruences=True, disable_equivalences=True):
+    @classmethod
+    def solve_supported_equivalence(cls, expr, solution):
+        # Given a supported equivalence, solve it explicitly by inverting Dim ops.
+        # TODO(avik): maybe do this with sympy in the future
```
@lezcano I feel like we've got another copy of this logic lying around somewhere?

lezcano replied:
Yep, this is a simpler version fo `try_solve`.

avikchaudhuri replied:
Hah yeah that's what jumped out to me today. (I'm sure `try_solve` does more.)

avikchaudhuri replied:
See above. Btw I was worried a bit about the limit of trials in try_solve but I realized that because we have linear expressions (of the form Ax + B) we only really ever need 2 trials, and the current limit is 5.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1560,7 +1610,7 @@ def relation_with_digit(expr, op, digit):
                 if t == s:
                     continue
-                left, op, right = t.split(" ")
+                left, op, right = t.split(" ", 2)
```
🤔

avikchaudhuri replied:
lol yeah we expect `symbol == expr` here, so 2 blanks. Will add comment.

avikchaudhuri replied:
Using re split now.

ezyang commented on torch/export/dynamic_shapes.py:
```diff
@@ -38,6 +42,64 @@ def readable(name, min_, max_):
             return f"Dim('{name}', min={min_})"
         return f"Dim('{name}', min={min_}, max={max_})"
 
+    def __add__(cls, other):
```
This is a preexisting thing, but why is a metaclass needed here?

ezyang replied:
Oh, is it so that you can put these inside type signatures? It's pretty hard to tell from this file alone because there's very little documentation on Dim

avikchaudhuri replied:
Yeah, that whole thing is a needless complication at this point, because we're not really doing that. `Dim` would be simpler as a regular class. In fact I would love to back it with a `sympy.expr` over a user-named symbol, no need for the lambda I think.

ezyang replied:
So, re having the symbols be a sympy.Expr, I am kind of ambivalent. I think it is OK as an API concept, but you're still going to do another translation layer under the hood to actually get it into ShapeEnv, because at the time you're assembling Dims we don't even have a ShapeEnv yet. The main place where you might screw yourself over is when Sympy made a bad API decision (like `a == b` being structural equality) and now you can't undo it without ... redoing it all yourself.

avikchaudhuri replied:
I'm not planning to do anything here in this PR.

ezyang commented on torch/export/_trace.py:
```diff
@@ -652,9 +652,16 @@ def forward(self, *args, **kwargs):
+                ep_non_strict.gm,
+            )
+        except (ConstraintViolationError, ValueRangeError) as e:
+            raise UserError(UserErrorType.CONSTRAINT_VIOLATION, str(e))  # noqa: TRY200
```
mfff, I'm not a huge fan of suppressing the inner exception here haha, especially when I'm debugging problems I often need the inner backtrace. Consider with_traceback instead of chaining if you think chaining is ugly

avikchaudhuri replied:
OK, will TAL.

avikchaudhuri replied:
Not planning to do anything here in this PR.

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2998,19 +3066,37 @@ def hint(s):
+                            f"the values of {self._debug_name(src)} = {src.name()} by "
+                            f"{self._debug_name(source)} = {expr.subs({symbol: sympy.Symbol(self._debug_name(src))})}."
+                        )
+                        record_constraint_violation(equalities_inputs.warn_only, self._debug_name(source), msg)
```
I think I'm just being lazy here. I should implement some transitive chaining of linear constraints. This *should not* be difficult.

avikchaudhuri replied:
Actually I might have a proof sketch when the root is a phantom symbol because we do some normalization already; we transitively propagate all derived dims based off such a symbol.

But for something that depends on an input source, here's a repro that highlights the lack of transitive equality reasoning you mention:

```
    class Foo(torch.nn.Module):
        def forward(self, x, y, z):
            return x + y[1:] + z[2:]

    foo = Foo()

    u, v, w = torch.randn(5), torch.randn(6), torch.randn(7)
    ep = torch.export.export(
        foo,
        (u, v, w),
        dynamic_shapes=({0: dimx}, {0: dimy}, {0: dimz}),
        strict=strict,
    )
    print(ep)
```

Dumb error follows!

```
torch._dynamo.exc.UserError: Constraints violated ((dimy+1), dimy)! For more information, run with TORCH_LOGS="+dynamic".
  - The values of dimy = L['y'].size()[0] must always be related to the values of (dimy-1) = L['x'].size()[0] by dimy = (dimy-1) + 1.
  - The values of (dimy+1) = L['z'].size()[0] must always be related to the values of (dimy-1) = L['x'].size()[0] by (dimy+1) = (dimy-1) + 2.

Suggested fixes:
  (dimy+1) = (dimy-1) + 2
  dimy = (dimy-1) + 1
```

After I fix this I will try to articulate why this can't happen with phantom symbols or come up with another repro for that and fix it.

avikchaudhuri replied:
This should now be fixed.

albanD commented on test/allowlist_for_publicAPI.json:
```diff
@@ -1987,7 +1987,7 @@
     "DimDynamic",
     "DynamicDimConstraintPrinter",
-    "EqualityConstraint",
+    "EqualityConstraints",
```
Adding or updating entries in this file is not ok. Please fix the public API of the module that you are using as suggested in the error message from test_public_bindings!

avikchaudhuri replied:
There is no error message from test_public_bindings (ran `buck2 test '@fbcode//mode/opt' fbcode//caffe2/test:public_bindings` to confirm). I was renaming `EqualityConstraint` to `EqualityConstraints` throughout. In any case this isn't really supposed to be a public API.

avikchaudhuri replied:
@albanD I could make `EqualityConstraints` private by adding `_`. However there seem to be many other things like it in this file we could also make private. cc @ezyang 

I prefer consistency of naming, so I'm torn between leaving this as is and making more things private. (Of course I could just do this for `EqualityConstraints` and move on, but I feel this is more process than needed in this PR.)

albanD replied:
A PR either making them all nicely public or nicely private would be good yes!

avikchaudhuri replied:
OK to leave this as is in this PR, then?

albanD replied:
It's best if you can do that other PR first and then rebase this on top since the general policy is that this allowlist is remove-only.

avikchaudhuri replied:
Eh, in that case I'd rather just back out the renaming and land this PR first.

avikchaudhuri commented:
@ezyang @lezcano this is ready for another review. Thanks!

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,71 @@ def render(self, source: Source):
+    form and so the problem reduces to symbolic expression equality.)
     """
     source_pairs: List[Tuple[Source, Source]]
+    derived_equalities: List[Tuple[Source, Union[Source, sympy.Symbol], Callable]]
```
so here Callable is a sympy.Expr to Expr function?

avikchaudhuri replied:
Yes, will add more explicit annot here.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,71 @@ def render(self, source: Source):
     """
     source_pairs: List[Tuple[Source, Source]]
+    derived_equalities: List[Tuple[Source, Union[Source, sympy.Symbol], Callable]]
+    phantom_symbols: Dict[str, sympy.Symbol]
```
What exactly is the string here?

avikchaudhuri replied:
It is the name of a phantom symbol. When building `EqualityConstraints`, the `phantom_symbols` field is populated by `dynamic_shapes.process_equalities`, where we use the key to point to sympy symbols created with the corresponding names.

When using `EqualityConstraints` in symbolic_shapes, the key doesn't really matter. So we could technically just pass in the values of the dict here and change the type to `List[sympy.Symbol]`.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,71 @@ def render(self, source: Source):
 DimConstraint = Union[StrictMinMaxConstraint, RelaxedUnspecConstraint, None]
 
 @dataclass(frozen=True)
```
This appears to be frozen but not actually (because the union find is mutable and stored here). It would be nice to explain exactly in what sense this is frozen.

avikchaudhuri replied:
It is frozen after post_init, i.e., once you pass in the public fields and we're done with initialization (which still doesn't touch the public fields, but does populate a couple of internal fields `_parents` and `_defs`), none of the fields are expected to change by code (even though the fields themselves are lists and dicts which are technically mutable).

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,71 @@ def render(self, source: Source):
+        # self._defs is a map from input sources to "canonical" symbolic expressions,
+        # i.e., unary expressions with symbols that corresponds to regular Dims (i.e.,
+        # not derived Dims)
+        object.__setattr__(self, "_defs", {})
```
type annots for these would be nice!

avikchaudhuri replied:
Will do.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,71 @@ def render(self, source: Source):
+            # preprocess into a transitively-closed map
+            # NOTE(avik): we reuse the union-find forest for canonicalizing input sources
+            if isinstance(root, sympy.Symbol):
+                self._defs[self._find(source)] = fn(root)
```
so is the source keyed on here the "canonical" source? When unification occurs, it's not clear to me how how these  keys also get correspondingly updated

ezyang replied:
Oh, maybe the point is that once this data structure is built we cannot learn more equalities, so the union find is fully baked at this point.

avikchaudhuri replied:
Yes exactly.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,73 @@ def render(self, source: Source):
     def _find(self, source):
+        # chase edges to find the root of this equivalence class
         if source in self._parents:
             return self._find(self._parents[source])
```
This is a preexisting condition, but you don't seem to path compress here?

avikchaudhuri replied:
No path compression or rank-based union. The number of things in this forest is typically small so I never bothered.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,73 @@ def render(self, source: Source):
+        for source, root, fn in self.derived_equalities:
+            # preprocess into a transitively-closed map
+            # NOTE(avik): we reuse the union-find forest for canonicalizing input sources
+            if isinstance(root, sympy.Symbol):
```
This is displeasingly asymmetric. What if we made phantom sources 🤔

avikchaudhuri replied:
We do have a fake (`ConstantSource`) source to play with. See `dynamic_shapes._process_equalities` for where we use `shape_env.create_symbol`. Would that be better?

ezyang replied:
It's not a big deal, it just took me along time to fully understand the new algorithm haha

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -800,29 +800,73 @@ def render(self, source: Source):
+    a given expression over another input source.
+    (NOTE: In contrast, it is easy to decide whether an input source is transitively equal
+    to a given expression over a phantom symbol; such expressions are already in canonical
+    form and so the problem reduces to symbolic expression equality.)
```
An example feels like it would help here. Maybe even the counterexample you gave earlier.

```
Suppose we have [a = b, b = c], [d = c + 1, e = d - 1]

We first construct a union find with source_pairs, giving us

_parents = {a: a, b: a, c: a}

Then, we compute canonical symbolic expressions, recursively applying unary expressions until we bottom out

_defs = {d: c + 1, e: (c + 1) - 1 aka c}
```

avikchaudhuri replied:
Good idea.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1463,6 +1516,18 @@ def solve(self, disable_congruences=True, disable_equivalences=True):
+    @classmethod
+    def _is_supported_equivalence(cls, expr):
+        # Currently supported Dim ops are linear expressions with integer coefficients.
+        # So check that expr only contains +, *, ints, and a single occurrence of a symbol
```
For future extenders, it would be helpful to explain why only linear expressions are supported

avikchaudhuri replied:
Agreed.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2804,22 +2871,49 @@ def is_dim(src):
+                expr2_ = fn(expr2)
+                # Check whether given input shape values satisfy a specified equation s = fn(s').
+                # - Raise when the equation was violated by the given input shape values.
+                # - Otherwise issue a guard to constrain them.
```
This is ever so slightly goofy, because we would prefer to issue the guard ASAP (to reduce the number of free symbols floating around), but it's not a big deal

ezyang reviewed (approved):
The updated algorithm is slick. Thanks!

lezcano reviewed (commented):
Just left a couple comments. As we discussed, all this logic could do with a big clean-up sooner than later. 

In particular, we should seriously streamline the class system here. One example: `_PhantomRoot` and `_ConstraintTarget` should be merged into one class, where `is_phantom` is given by `t_id is None`, for example.

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -394,17 +569,86 @@ def tree_zip(combined_args, dynamic_shapes):
+        def root_value():
+            # given tensor.shape[i] is the value of dim = fn(root),
+            # find the value of root
+            symbol = sympy.Symbol(dim.root.__name__, integer=True)
```
I'll clamp the bounds. We haven't yet moved the bounds to `>= 0` by the way, that's a TODO (somebody else is working on a PR for that).

avikchaudhuri replied:
To be clear, I'm not going to use `nonnegative` (but instead enforce it via bounds) because we have some work TODO on serde to carry assumptions on all symbols in shape env, and I don't want to add assumptions that cannot be maintained through serde. (The only option is to also blanket-use `nonnegative` there right now, which is not correct for unbacked symints. I have a comment on that in this PR near serde logic.)

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -38,6 +42,76 @@ def readable(name, min_, max_):
+        assert type(other) is int, f"Expected int, got {type(other)}"
+        return cls._derive(lambda x: x - other)
+
+    def __rsub__(cls, other):
+        raise NotImplementedError("non-monotonic operation not supported")
```
Actually I think the technical term is "monotonically increasing." x -> -x is monotonically decreasing. We also have that a = b -> f(a) = f(b). I'll add the equation and change the proper terminology.

lezcano replied:
So, if `a = b` you always have `f(a) = f(b)` :p
If you want `a <= b -> f(a) <= f(b)`, that's simply "increasing".
Monotonically increasing is redundant. Monotonically = increasing or decreasing.
If you want `a <= b -> f(a) <= f(b)` then you need to add zero as an allowed value in `__mul__`.

At any rate, yes, please add the formula as it'll make things much clearer.

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -394,17 +569,86 @@ def tree_zip(combined_args, dynamic_shapes):
+                    root_constraint.dim,
+                )
+            elif dim_root.__name__ not in phantom_roots:
+                # create a phantom root
```
There is only one place in dynamic_shapes right now where we create `_PhantomRoot`.

lezcano replied:
That's right... I don't know what I was referring to... Sorry for the noise, will post if I remember.

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -38,6 +42,76 @@ def readable(name, min_, max_):
+
+    def __mul__(cls, other):
+        # e.g., dim * 2
+        assert type(other) is int, f"Expected int, got {type(other)}"
```
I'll merge the two messages into one `NotImplementedError`.

avikchaudhuri committed:
derived dim (#118729)

Summary:
With the current `Dim`-based dynamic shapes API for export, one can express that shapes of different input shapes must be equal by reusing the same `Dim`. However, non-trivial relationships between such input shapes cannot be expressed.

Recently we are seeing more and more examples of code that require this additional expressibility, e.g., where a pair of shapes might differ by one, or a shape might be double another (or simply even).

This PR introduces the concept of a "derived" `Dim`, i.e., a linear arithmetic expression over a `Dim`. By using a combination of `Dim`s and derived `Dim`s to specify input shapes, the desired relationships can be expressed naturally. E.g., a pair of shapes might be `dim` and `dim + 1`, or `dim` and `2*dim`, or even `2*dim` and `dim + 1`.

We extend the current infrastructure that translates `Dim`s to deprecated `dynamic_dim`-based constraints to work with derived `Dim`s. As usual, we raise constraint violation errors when shape guards cannot be verified given a dynamic shapes spec; suggest fixes; and raise runtime errors when future inputs violate the spec.

Importantly, some guards that used to cause forced specializations in the constraint solver because they were deemed "too complex" now do not do so, because they can now be specified as constraints. Since this was what motivated the introduction of a `disable_constraint_solver` flag to some internal APIs, we may not need that flag any more.

Note that shapes of placeholders in exported programs can now contain symbolic expressions and not just symbols.

bypass-github-pytorch-ci-checks

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Pull Request resolved: https://github.com/pytorch/pytorch/pull/118729

Reviewed By: angelayi

Differential Revision: D53254587

Pulled By: avikchaudhuri
